### PR TITLE
throttle diskcaching writes [perf]

### PR DESCRIPTION
**Pull request by thundergolfer**
**TLDR:** On `cold_start.rs` this change shows a 25-30% performance improvement.

----

On the local `cold_start.rs` benchmark, a plain, non-caching `Remote` provider gets 400MiB/s throughput. The disk and mem caching `Cache` provider gets about 220-230 MiB/s. This is a large cold-start penalty to pay for subsequent runs to be fast. 

If you simply remove the diskcaching writes — I left the provider pointlessly checking for diskcache hits — then the `Cache` provider, which is now just mem-caching, jumps up in throughput to ~380 MiB/s. So the writes themselves seem to be very costly, even though they're _supposed to be non-blocking and highly concurrent._

These `spawn_blocking` writes are non-blocking, but they're clearly tanking performance. 

The Tokio runtime has a finite amount (512) of blocking threads it will spawn. This is a shared resource, and other aspects of the worker application (incl. networking) use these blocking threads, for example DNS lookups. If I purposely tank the diskcaching performance by adding a sync `std:fs::sleep` of 10s, the worker slows to a crawl as all the Tokio runtime's blocking threads become occupied sleeping.

I did some profiling with `perf` and could see that increased disk caching resulted in large numbers of `tokio-runtime-w` blocking working threads, and it you can see that some are doing network stuff and most are doing file IO.

<kbd>
image
</kbd>

The mostly orange (kernel samples) threads are the file IO threads. 

But this profiling doesn't show me why the Tokio runtime is slowed down by needing to spawn and manage these blocking threads. On 2GB reads the current worker+blobnet code doesn't get close to hitting the max 512 blocking threads. Maybe https://docs.rs/tokio-console/latest/tokio_console/ will be a better tool understand where the slowdown is coming from. 

**The change**

By throttling diskcaching writes with a Semaphore, something that's called out in the `spawn_blocking` docstring, the runtime spawns fewer blocking threads for file IO and the overall cold FUSE read speeds up. 

* Throughput on the `cold_start.rs` benchmark rises from 220-230MiB/s to 300-310MiB/s. 👍 
* I can also observe this speed up just running Modal apps with large cold reads locally. A 2GB read goes from 8-9s down to 5-5.5s. 👍 

I can observe that on a 2GB cold file read in the worker, around 300-400MB of disk cache writing happens _after_ the Python process has finished reading the file. If the Python process remains alive after the read has completed, so does the FUSE filesystem and so the remaining data is written to disk. If the Python process is immediately killed then these pending diskcache writes are cancelled; they don't block the `umount2` cleanup operation.

This implementation with a tokio `Semaphore` could probably be improved on. It's what I came up with first.

**Next steps**

* Still don't understand why the worker is so slowed down by spawning these blocking threads for file I/O. 
* Related to the above, the `Cached` provider is still markedly slower than non-caching `Remote`. Could we reduce the impact of file I/O further?
* Tuning the semaphore count. Should it be `4`? Dropping to `2` doesn't seem to improve performance on `cold_start.rs`.

### CODE REVIEW

File: blobnet/src/provider.rs [hunk 1]
```diff
@@ -452,6 +453,7 @@ impl<P> Cached<P> {
                 pending_requests: Default::default(),
                 dir: dir.as_ref().to_owned(), // File system cache
                 pagesize,
+                diskcache_semaphore: tokio::sync::Semaphore::new(4).into(),
```
Comment (ekzhang) [inline blobnet/src/provider.rs:456]: Is it intended to be limited to only `4` threads?

**Review by ekzhang — APPROVED** [review]
Great improvement!

**Comment by erikbern** [comment]
> I can also observe this speed up just running Modal apps with large cold reads locally. A 2GB read goes from 8-9s down to 5-5.5s. 👍

Wow, that's huge!! A bit unsettling that we don't fully understand _why_ this is faster, but I'm sure we can figure it out later.

File: blobnet/src/provider.rs (hunk 1 above)
Comment (thundergolfer) [reply blobnet/src/provider.rs:456]: Yes, intended, but this value wasn't chosen scientifically. I initially tried 128, then 32, then 4, then 2.

Comment (ekzhang) [reply blobnet/src/provider.rs:456]: I'm a bit worried that this number is small enough that it may cause a lot of disk operations to build up, stuck at the semaphore point, instead of being executed immediately.

Comment (thundergolfer) [reply blobnet/src/provider.rs:456]: What's the worst outcome of a backlog? It can mean lost writes to the cache dir, which will slow down subsequent reads, but is there something worse?

Comment (ekzhang) [reply blobnet/src/provider.rs:456]: A backlog will increase RAM usage by at least 2 MiB per thread

Comment (thundergolfer) [reply blobnet/src/provider.rs:456]: Oh of course. It would be great to be able to reason about the relation between the semaphore number and disk write throughput. Any investigative direction you can suggest?

Comment (ekzhang) [reply blobnet/src/provider.rs:456]: Maybe something like this?

```rust
use std::path::Path;
use std::sync::Arc;
use std::time::Instant;

use anyhow::Result;
use tokio::fs;
use tokio::sync::Semaphore;

#[tokio::main]
async fn main() -> Result<()> {
    let n = 2048; // 2 MiB * 2048 = 4 GiB

    let folder = Path::new("/tmp/.cc-test");
    fs::create_dir_all(folder).await?;

    for permits in [1, 2, 4, 8, 16, 32, 64] {
        let sema = Arc::new(Semaphore::new(permits));

        let start = Instant::now();
        for i in 0..n {
            let filename = folder.join(format!("{i}.txt"));
            let permit = sema.clone().acquire_owned().await?;
            tokio::spawn(async move {
                let _permit = permit;
                fs::write(filename, "a".repeat(1 << 21)).await.unwrap();
            });
        }
        println!("permits = {permits}, time = {:?}", start.elapsed());
    }

    Ok(())
}
```

Comment (thundergolfer) [reply blobnet/src/provider.rs:456]: As shown IRL, we needed an fsync to get proper results. Modified script:

```
use std::path::Path;
use std::sync::Arc;
use std::time::Instant;

use anyhow::Result;
use tokio::fs;
use tokio::sync::Semaphore;

#[tokio::main]
async fn main() -> Result<()> {
    let n = 2048; // 2 MiB * 2048 = 4 GiB

    let folder = Path::new("/tmp/.cc-test");
    fs::create_dir_all(folder).await?;

    for permits in [1, 2, 4, 8, 16, 32, 64] {
        let sema = Arc::new(Semaphore::new(permits));

        let start = Instant::now();
        for i in 0..n {
            let filename = folder.join(format!("{i}.txt"));
            let permit = sema.clone().acquire_owned().await?;
            tokio::spawn(async move {
                let _permit = permit;
                tokio::task::spawn_blocking(move || {
                    let mut f = std::fs::File::create(filename).unwrap();
                    f.write("a".repeat(1 << 21).as_bytes()).unwrap();
                    f.sync_data().unwrap();
                    // std::fs::write(filename, "a".repeat(1 << 21)).unwrap();
                })
                .await
                .unwrap();
            });
        }
        println!("permits = {permits}, time = {:?}", start.elapsed());
    }

    Ok(())
}
```

**results:**

```
permits = 1, time = 15.791450323s
permits = 2, time = 16.582980209s
permits = 4, time = 16.603092725s
permits = 8, time = 16.47892605s
permits = 16, time = 16.450045673s
permits = 32, time = 16.337582942s
permits = 64, time = 16.370281234s
```

Comment (thundergolfer) [reply blobnet/src/provider.rs:456]: More interesting results on a g5.12xlarge, which has an SSD:

```
ubuntu@ip-172-31-2-29:~$ sudo ./concurr-writes
permits = 1, time = 7.266329923s
permits = 2, time = 6.251999261s
permits = 4, time = 4.320230757s
permits = 8, time = 3.212917571s
permits = 16, time = 3.142664171s
permits = 32, time = 3.499568409s
permits = 64, time = 3.495804387s
ubuntu@ip-172-31-2-29:~$ sudo ./concurr-writes
permits = 1, time = 8.597891941s
permits = 2, time = 6.292242716s
permits = 4, time = 4.346956102s
permits = 8, time = 3.19075752s
permits = 16, time = 3.218678721s
permits = 32, time = 3.574626054s
permits = 64, time = 3.465188719s
ubuntu@ip-172-31-2-29:~$
```

Comment (ekzhang) [reply blobnet/src/provider.rs:456]: Ahh, so it looks like the local SSD does have more throughput then! Good to know :D

Permits = 4 looks good I think.

Comment (luiscape) [reply blobnet/src/provider.rs:456]: I'm curious to try this same test on the fastest EBS drive AWS has: [io2 Block Express](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-volume-types.html).
